url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8/src/com/fuzzylite/Console.java

```diff
@@ -0,0 +1,623 @@
+                }
+            }
+        }
+
+        String outputFilename = options.get(KW_OUTPUT_FILE);
+        String outputFormat = options.get(KW_OUTPUT_FORMAT);
+        if (outputFormat == null || outputFormat.isEmpty()) {
+            if (outputFilename == null || outputFilename.isEmpty()) {
+                throw new RuntimeException("[format error] unspecified format of output");
+            } else {
+                int extensionIndex = outputFilename.lastIndexOf(".");
+                if (extensionIndex >= 0) {
+                    outputFormat = outputFilename.substring(extensionIndex + 1);
+                } else {
+                    throw new RuntimeException("[format error] unspecified format of output file");
+                }
+            }
+        }
+
+        Writer writer;
+        if (outputFilename == null || outputFilename.isEmpty()) {
+            writer = new StringWriter();
+        } else {
+            File outputFile = new File(outputFilename);
+            if (!outputFile.exists()) {
+                outputFile.createNewFile();
+            }
+            writer = new FileWriter(outputFile);
+        }
+        try {
+            process(textEngine.toString(), writer, inputFormat, outputFormat, options);
+        } catch (Exception ex) {
+            writer.close();
+            throw ex;
+        }
+        writer.flush();
+        if (outputFilename == null || outputFilename.isEmpty()) {
+            StringWriter stringWriter = (StringWriter) writer;
+            System.out.println(stringWriter.getBuffer().toString());
+        }
+        writer.close();
+    }
+
+    protected static void process(String input, Writer writer,
+            String inputFormat, String outputFormat, Map<String, String> options)
+            throws Exception {
+        Importer importer = null;
+        if ("fll".equals(inputFormat)) {
+            importer = new FllImporter();
+        } else if ("fcl".equals(inputFormat)) {
+            importer = new FclImporter();
+        } else if ("fis".equals(inputFormat)) {
+            importer = new FisImporter();
+        } else {
+            throw new RuntimeException(String.format(
+                    "[import error] format <%s> not supported", inputFormat));
+        }
+        Engine engine = importer.fromString(input);
+
+        if ("fld".equals(outputFormat)) {
+            FldExporter fldExporter = new FldExporter();
+            String filename = options.get(KW_DATA_INPUT);
+            if (filename != null) {
+                File dataFile = new File(filename);
+                if (!dataFile.exists()) {
+                    throw new RuntimeException("[export error] file <" + filename + "> "
+                            + "does not exist");
+                }
+                writer.write("#" + fldExporter.header(engine) + "\n");
+                BufferedReader reader = new BufferedReader(new FileReader(dataFile));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced at lines 275, 279. The resource is closed at line 297. There are other execution paths that do not contain closure statements. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).